src/components/coffeechat.ts

```diff
@@ -0,0 +1,392 @@
+  for (let i = 0; true; i++) {
+    //generate one week of matches, and updates match freq tables accordingly
+    const nextResults = stableMatch(userList, matched);
+    for (const pair of nextResults) {
+      matched[userList.get(pair[0])!][userList.get(pair[1])!]++;
+      matched[userList.get(pair[1])!][userList.get(pair[0])!]++;
+      newMatches[userList.get(pair[0])!][userList.get(pair[1])!]++;
+      newMatches[userList.get(pair[1])!][userList.get(pair[0])!]++;
+    }
+
+    //stops and returns if most recent match caused a duplicate matching to appear
+    if (hasDupe(newMatches, nextResults, userList)) {
+      return i;
+    }
+
+    //if no dupe, write this new match to results and continue generating
+    await writeFutureMatches(nextResults, i);
+  }
+};
+
+/*
+ * Get the ID's of users for which future matches was generated for
+ * Returns the above as a array of strings
+ */
+const getFutureMatchIDs = async (): Promise<string[]> => {
+  const db = await openDB();
+  return (
+    await db.all(
+      `SELECT first_user_id FROM coffee_future_matches UNION SELECT second_user_id FROM coffee_future_matches`
+    )
+  ).map((val) => val['first_user_id']);
+};
+
+//tests if generated matches still work for people wanting a match
+//and if unused generated matches still remain
+export const validateFutureMatches = async (client: Client): Promise<boolean> => {
+  const db = await openDB();
+
+  //check if the ID's of users CURRENTLY wanting coffee chats are the same as the
+  //users we have generated matches for
+  const activeMatchIDs = [...(await loadNotMatched(client)).keys()];
+  const generatedMatchIDs = await getFutureMatchIDs();
+
+  //checks if same amount of ID's on both sides, then checks that they are equal
+  if (activeMatchIDs.length != generatedMatchIDs.length) return false;
+  for (const key of activeMatchIDs) {
+    if (!generatedMatchIDs.includes(key)) return false;
+  }
+
+  //test if there exists week_id that is not finished
+  const test_id = await db.get(`SELECT * FROM coffee_week_status WHERE finished = 0`);
+  if (!test_id) return false;
+
+  return true;
+};
+
+/*
+ * Pulls the next round of pre-generated matches
+ * ONLY run this is you know there still exists at least one more round of valid matches
+ * Verify this by running validdateFutureMatches prior to this
+ * Return an array of matches (themselves an array of size 2); potential "single" person's ID will be returned in single, null otherwise
+ */
+export const getNextFutureMatch = async (client: Client): Promise<{ matches: string[][]; single: string | null }> => {
+  const db = await openDB();
+
+  //get matches for the smallest week_id that hasn't been used, then set the week_id to be used
+  const { week_id } = (await db.get(`SELECT min(week_id) AS week_id FROM coffee_week_status WHERE finished = 0;`)) as {
+    week_id: number;
+  };
+  const matches = ((await db.all(
+    `SELECT * FROM coffee_future_matches WHERE week_id = ${week_id};`
```

Comment:
do not use template strings for SQL queries, use the builtin parameter feature, this is prone to a SQL injection attack

Reply:
```
db.all('SELECT * FROM coffee_future_matches WHERE week_id = ?', week_id)
```